lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = MappingProxyType(
+    {
+        "_5f": "_",
+        "_40": "@",
+        "_2e": ".",
+        "_2d": "-",
+        "_5c": "\\",
+    }
+)
+
+
+def _systemctl(*args) -> None:
+    """Control systemd by via executed `systemctl ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if systemctl command fails.
+    """
+    cmd = ["systemctl", *args]
+    _logger.debug(f"systemd: Executing command {cmd}")
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        _logger.error(f"systemctl command failed: {e}")
+        raise
+
+
+_daemon_reload = functools.partial(_systemctl, "daemon-reload")
+_daemon_reload.__doc__ = "Reload systemd manager configuration."
+_start_service = functools.partial(_systemctl, "start")
+_start_service.__doc__ = "Start systemd service unit."
+_stop_service = functools.partial(_systemctl, "stop")
+_stop_service.__doc__ = "Stop systemd service unit."
+_enable_service = functools.partial(_systemctl, "enable")
+_enable_service.__doc__ = "Enable systemd service."
+_disable_service = functools.partial(_systemctl, "disable")
+_disable_service.__doc__ = "Disable systemd service."
+
+
+class ServiceStartedEvent(EventBase):
+    """Event emitted when service has started."""
+
+
+class ServiceStoppedEvent(EventBase):
+    """Event emitted when service has stopped."""
+
+
+class SystemdNotices:
+    """Observe systemd services on your machine base."""
+
+    def __init__(self, charm: CharmBase, services: Union[str, List[str]]) -> None:
+        """Instantiate systemd notices service."""
+        self._charm = charm
+        self._services = [services] if isinstance(services, str) else services
+        unit_name = self._charm.unit.name.replace("/", "-")
+        self._service_file = Path(f"/etc/systemd/system/juju-{unit_name}-systemd-notices.service")
+
+        _logger.debug(f"Attaching systemd notice events to charm {self._charm.__class__.__name__}")
+        for service in self._services:
+            self._charm.on.define_event(f"service_{service}_started", ServiceStartedEvent)
+            self._charm.on.define_event(f"service_{service}_stopped", ServiceStoppedEvent)
+
+    def subscribe(self) -> None:
+        """Subscribe charmed operator to observe status of systemd services."""
+        _logger.debug(f"Generating systemd notice hooks for {self._services}")
+        for service in self._services:
+            Path(f"hooks/service-{service}-started").symlink_to(f"{Path.cwd()}/dispatch")
+            Path(f"hooks/service-{service}-stopped").symlink_to(f"{Path.cwd()}/dispatch")
```

Comment:
It seems weird to use the current working directory here (global state). Is there a way to calculate this from the charm directory? `Framework` has a `charm_dir` attribute, for example. Similar elsewhere.

Reply:
Hmm... it makes sense to use the `charm_dir` attribute in the public charm API part of the charm lib, but we cannot use it in the daemon part of the charm lib. Is there a particular concern with using `Path.cwd()` in the daemon bits of this charm lib? While we might have to worry about someone changing the current working directory inside their charm before setting up the `juju-systemd-notices` daemon, but the in the daemon bits we know that we will be in the correct directory since we set the WorkingDirectory in the service file for the daemon.

Reply:
Ah yeah, fair enough -- when I wrote this comment I wasn't thinking about the daemon part. I think that's fine: let's use `charm_dir` in the API part but `Path.cmd()` is fine in the daemon part.

Reply:
Replaced `Path.cwd()` with `self._charm.framework.charm_dir` in the public charm API in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/a61cf98ddfaf54f73b3c56c9182208052ca01032